src/common.c

```diff
@@ -707,60 +709,191 @@ int fi_epoll_create(struct fi_epoll **ep)
-			     sizeof(*ep->fds) + sizeof(*ep->context));
-		if (!fds)
-			return -FI_ENOMEM;
-
-		ep->size += 64;
-		contexts = fds + ep->size;
+	struct fi_epoll_work_item *item;
 
-		memcpy(fds, ep->fds, ep->nfds * sizeof(*ep->fds));
-		memcpy(contexts, ep->context, ep->nfds * sizeof(*ep->context));
-		free(ep->fds);
-		ep->fds = fds;
-		ep->context = contexts;
-	}
+	item = calloc(1,sizeof(*item));
+	if (!item)
+		return -FI_ENOMEM;
 
-	ep->fds[ep->nfds].fd = fd;
-	ep->fds[ep->nfds].events = events;
-	ep->context[ep->nfds++] = context;
+	item->fd = fd;
+	item->events = events;
+	item->context = context;
+	item->type = FI_EPOLL_ADD;
+	fastlock_acquire(&ep->lock);
+	slist_insert_tail(&item->entry, &ep->work_item_list);
 	fd_signal_set(&ep->signal);
+	fastlock_release(&ep->lock);
 	return 0;
 }
 
 int fi_epoll_mod(struct fi_epoll *ep, int fd, uint32_t events, void *context)
 {
+	struct slist_entry *entry, *prev;
+	struct fi_epoll_work_item *item, *mod_item;
 	int i;
 
-	for (i = 0; i < ep->nfds; i++) {
-		if (ep->fds[i].fd == fd) {
-			ep->fds[i].events = events;
-			ep->context[i] = context;
-			fd_signal_set(&ep->signal);
-			return 0;
-		}
+	mod_item = calloc(1,sizeof(*mod_item));
+	if (!mod_item)
+		return -FI_ENOMEM;
+
+	fastlock_acquire(&ep->lock);
+	slist_foreach(&ep->work_item_list, entry, prev) {
+		item = container_of(entry, struct fi_epoll_work_item, entry);
+		if (item->fd != fd)
+			continue;
+
+		item->events = events;
+		item->context = context;
+		fd_signal_set(&ep->signal);
+		fastlock_release(&ep->lock);
+		free(mod_item);
+		return 0;
+	}
+
+	for (i = 1; i < ep->nfds; i++) {
+		if (ep->fds[i].fd != fd)
+			continue;
+
+		mod_item->fd = fd;
+		mod_item->events = events;
+		mod_item->context = context;
+		mod_item->type = FI_EPOLL_MOD;
+		slist_insert_tail(&mod_item->entry,
+				  &ep->work_item_list);
+		fd_signal_set(&ep->signal);
+		fastlock_release(&ep->lock);
+		return 0;
   	}
+	fastlock_release(&ep->lock);
+	free(mod_item);
 	return -FI_EINVAL;
 }
 
 int fi_epoll_del(struct fi_epoll *ep, int fd)
 {
+	struct slist_entry *entry, *prev;
+	struct fi_epoll_work_item *item, *del_item;
 	int i;
 
-	for (i = 0; i < ep->nfds; i++) {
+	del_item = calloc(1,sizeof(*item));
+	if (!del_item)
+		return -FI_ENOMEM;
+
+	del_item->fd = fd;
+	del_item->type = FI_EPOLL_DEL;
+
+	fastlock_acquire(&ep->lock);
+	slist_foreach(&ep->work_item_list, entry, prev) {
+		item = container_of(entry, struct fi_epoll_work_item,
+				    entry);
+		if (item->fd == fd &&
+		    (item->type == FI_EPOLL_ADD)) {
+			goto out;
+		}
+	}
+
+	for (i = 1; i < ep->nfds; i++) {
 		if (ep->fds[i].fd == fd) {
-			ep->fds[i].fd = ep->fds[ep->nfds - 1].fd;
-			ep->context[i] = ep->context[--ep->nfds];
-			fd_signal_set(&ep->signal);
-			return 0;
+			goto out;
 		}
   	}
+	fastlock_release(&ep->lock);
+	free(del_item);
 	return -FI_EINVAL;
+out:
+	slist_insert_tail(&del_item->entry, &ep->work_item_list);
+	fd_signal_set(&ep->signal);
+	fastlock_release(&ep->lock);
+	return 0;
```

Comment:
Same comment as for mod -- I think we can just queue the delete call and let the handling code deal with things.